File: src/com/study/algorithm/algorithms/algorithmsJAVA/岛屿数量.java
package com.study.algorithm.algorithms.algorithmsJAVA;

import org.junit.Test;

@Deprecated
public class 岛屿数量 {
    @Test
    public void 岛屿数量() {
        // 求 1+2+...+n
        char[][] chars =
                new char[][] {
                        {'1', '1', '1', '1', '0'},
                        {'1', '1', '0', '1', '0'},
                        {'1', '1', '0', '0', '0'},
                        {'0', '0', '0', '0', '0'}
                };

        System.out.println("岛屿数量:" + numIslands(chars));
    }
    /**
     * 给你一个由 '1'（陆地）和 '0'（水）组成的的二维网格，请你计算网格中岛屿的数量。
     *
     * <p>岛屿总是被水包围，并且每座岛屿只能由水平方向和/或竖直方向上相邻的陆地连接形成。
     *
     * <p>此外，你可以假设该网格的四条边均被水包围。
     *
     * <p>示例 1:
     *
     * <p>输入: 11110 11010 11000 00000 输出: 1 示例 2:
     *
     * <p>输入: 11000 11000 00100 00011 输出: 3 解释: 每座岛屿只能由水平和/或竖直方向上相邻的陆地连接而成。
     *
     * <p>来源：力扣（LeetCode） 链接：https://leetcode-cn.com/problems/number-of-islands
     * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
     */
    public int numIslands(char[][] grid) {
        /*todo 未完成*/
        if (grid == null || grid.length == 0) {
            return 0;
        }

        int num = 0;
        int row = grid.length, col = grid[0].length, i, j;

        for (i = 0; i < row; i++) {
            for (j = 0; j < col; j++) {
                if (grid[i][j] == '1') {
                    if ((i == 0 || grid[i - 1][j] == '0')
                            && (row == 1 || (i < row - 1 && grid[i + 1][j] == '0'))
                            && (j == 0 || grid[i][j - 1] == '0')
                            && (col == 1 || (j < col - 1 && grid[i][j + 1] == '0'))) {
                        num++;
                    }
                }
            }
        }

        return num;
    }
}

File: src/com/study/algorithm/algorithms/algorithmsKT/岛屿数量.kt
package com.study.algorithm.algorithms.algorithmsKT

//给你一个由 '1'（陆地）和 '0'（水）组成的的二维网格，请你计算网格中岛屿的数量。
//
//岛屿总是被水包围，并且每座岛屿只能由水平方向和/或竖直方向上相邻的陆地连接形成。
//
//此外，你可以假设该网格的四条边均被水包围。
//
// 
//
//示例 1：
//
//输入：grid = [
//  ["1","1","1","1","0"],
//  ["1","1","0","1","0"],
//  ["1","1","0","0","0"],
//  ["0","0","0","0","0"]
//]
//输出：1
//示例 2：
//
//输入：grid = [
//  ["1","1","0","0","0"],
//  ["1","1","0","0","0"],
//  ["0","0","1","0","0"],
//  ["0","0","0","1","1"]
//]
//输出：3
// 
//
//提示：
//
//m == grid.length
//n == grid[i].length
//1 <= m, n <= 300
//grid[i][j] 的值为 '0' 或 '1'
//
//来源：力扣（LeetCode）
//链接：https://leetcode.cn/problems/number-of-islands
//著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
class 岛屿数量 {

    var res = 0
    fun numIslands(grid: Array<CharArray>): Int {
        //深度优先搜索
        //遍历矩阵,遇到'1'时进行深度优先搜索,并用一个布尔值的矩阵记录改格子是否被遍历过
        //如果在深度优先搜索中,遇到'1',将改网格记录为已查找,下次遇到时直接跳过
        //每次遇到新的'1'的格子,并且没被记录过,则说明遇到一个新的"岛屿",将数量+1
        val visit = Array(grid.size) {
            BooleanArray(grid[0].size)
        }
        for (r in grid.indices) {
            for (c in grid[0].indices) {
                if (!visit[r][c] && grid[r][c] == '1') {
                    res++
                    dfs(r, c, grid, visit)
                }
            }
        }
        return res
    }

    fun dfs(row: Int, col: Int, grid: Array<CharArray>, visit: Array<BooleanArray>) {
        if (!check(row, col, grid)) {
            return
        }
        if (visit[row][col]) {
            return
        }
        if (grid[row][col] != '1') {
            return
        }
        visit[row][col] = true
        dfs(row - 1, col, grid, visit)
        dfs(row, col + 1, grid, visit)
        dfs(row + 1, col, grid, visit)
        dfs(row, col - 1, grid, visit)
    }

    fun check(row: Int, col: Int, grid: Array<CharArray>): Boolean {
        if (row < 0 || col < 0 || row == grid.size || col == grid[0].size) {
            return false
        }
        return true
    }
}